tests/unit/s2n_self_talk_client_hello_cb_test.c

```diff
@@ -211,11 +231,38 @@ int s2n_negotiate_nonblocking_ch_cb(struct s2n_connection *conn,
     return s2n_negotiate(conn, &blocked);
 }
 
+int s2n_negotiate_nonblocking_poll(struct s2n_connection *conn,
+    struct client_hello_context *ch_ctx)
+{
+    s2n_blocked_status blocked;
+    EXPECT_NOT_NULL(conn);
+
+    EXPECT_EQUAL(ch_ctx->invoked, 0);
+
+    /* negotiate handshake, we should pause after the nonblocking callback is invoked */
+    EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate(conn, &blocked), S2N_ERR_ASYNC_BLOCKED);
+    EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_APPLICATION_INPUT);
+    EXPECT_EQUAL(ch_ctx->invoked, 1);
+
+    /* unless explicitly unblocked we should stay paused */
+    EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate(conn, &blocked), S2N_ERR_ASYNC_BLOCKED);
+    EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_APPLICATION_INPUT);
+    EXPECT_EQUAL(ch_ctx->invoked, 2);
+
+    ch_ctx->mark_done = true;
+
+    /* Expect the callback to complete after 2nd iteration */
+    EXPECT_SUCCESS(s2n_negotiate(conn, &blocked));
+    EXPECT_EQUAL(ch_ctx->invoked, 3);
+
+    return S2N_SUCCESS;
+}
+
 int s2n_negotiate_blocking_ch_cb(struct s2n_connection *conn, struct client_hello_context *ch_ctx)
 {
     s2n_blocked_status blocked;
     EXPECT_NOT_NULL(conn);
-    
+
```

Comment:
If you're going to be working with existing s2n-tls code, you might want to check if your IDE can avoid making whitespace changes to code that you don't touch. Otherwise your diffs might get weird :\

Reply:
Oh yes it does work! Although now I'm guessing I won't know if people add wonky whitespace back xD